network/gossip/libp2p/topology/topicBasedTopology.go

```diff
@@ -6,66 +6,172 @@ import (
+	myChannelIDs := t.subMngr.GetChannelIDs()
+	if len(myChannelIDs) == 0 {
+		// no subscribed channel id, hence skip topology creation
+		// we do not return an error at this state as invocation of MakeTopology may happen before
+		// node subscribing to all its channels.
+		return flow.IdentityList{}, nil
+	}
 
-	if _, ok := engine.IsClusterChannelID(topic); ok {
-		// extracts cluster peer ids to which the node belongs to.
-		clusterPeers, err := t.clusterPeers()
+	// finds all interacting roles with this node
+	myInteractingRoles := flow.RoleList{}
+	for _, myChannel := range myChannelIDs {
+		roles, ok := engine.RolesByChannelID(myChannel)
+		if !ok {
+			return nil, fmt.Errorf("could not extract roles for channel: %s", myChannel)
+		}
+		myInteractingRoles = myInteractingRoles.Union(roles)
+	}
+
+	// builds a connected component per role this node interact with,
+	var myFanout flow.IdentityList
+	for _, role := range myInteractingRoles {
+		if role == flow.RoleCollection {
+			// we do not build connected component for collection nodes based on their role
+			// rather we build it based on their cluster identity in the next step.
+			continue
+		}
+		roleFanout, err := t.subsetRole(ids, nil, flow.RoleList{role})
 		if err != nil {
-			return nil, fmt.Errorf("failed to find cluster peers for node %s: %w", t.me.String(), err)
+			return nil, fmt.Errorf("failed to derive list of peer nodes to connect for role %s: %w", role, err)
 		}
+		myFanout = myFanout.Union(roleFanout)
+	}
 
-		subscribers = clusterPeers
-	} else {
-		// not a cluster-based topic.
-		//
-		// extracts flow roles subscribed to topic.
-		roles, ok := engine.RolesByChannelID(topic)
-		if !ok {
-			return nil, fmt.Errorf("unknown topic with no subscribed roles: %s", topic)
+	// stitches the role-based components that subscribed to the same channel id together.
+	for _, myChannel := range myChannelIDs {
+		shouldHave := myFanout.Copy()
+
+		topicFanout, err := t.subsetChannel(ids, shouldHave, myChannel)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive list of peer nodes to connect for topic %s: %w", myChannel, err)
 		}
+		myFanout = myFanout.Union(topicFanout)
+	}
 
-		// extract ids of subscribers to the topic
-		subscribers = ids.Filter(filter.HasRole(roles...))
+	if len(myFanout) == 0 {
+		return nil, fmt.Errorf("topology size reached zero")
 	}
+	return myFanout, nil
+}
+
+// subsetChannel returns a random subset of the identity list that is passed. `shouldHave` represents set of
+// identities that should be included in the returned subset.
+// Returned identities should all subscribed to the specified `channel`.
+// Note: this method should not include identity of its executor.
+func (t *TopicBasedTopology) subsetChannel(ids flow.IdentityList, shouldHave flow.IdentityList,
+	channel string) (flow.IdentityList, error) {
+	if _, ok := engine.IsClusterChannelID(channel); ok {
+		return t.clusterChannelHandler(ids, shouldHave)
+	} else {
```

Comment:
```suggestion
	}
```

Reply:
we can remove the `else`

Reply:
fixed in https://github.com/onflow/flow-go/pull/153/commits/d957167069a961baed6e9f09f414ab8b990b4df2.